lib/block-transform-stream.ts

```diff
@@ -16,60 +16,92 @@
 
 import { Transform } from 'readable-stream';
 
+import { AlignedReadableState } from './aligned-lockable-buffer';
+import { CHUNK_SIZE } from './constants';
+
 export class BlockTransformStream extends Transform {
 	public bytesRead = 0;
 	public bytesWritten = 0;
-	private _buffers: Buffer[] = [];
-	private _bytes = 0;
+	private chunkSize: number;
+	private alignedReadableState: AlignedReadableState;
+	private inputBuffers: Buffer[] = [];
+	private inputBytes = 0;
 
-	constructor(private chunkSize: number) {
-		super();
+	constructor({
+		chunkSize,
+		alignment,
+		numBuffers = 2,
+	}: {
+		chunkSize: number;
+		alignment: number;
+		numBuffers?: number;
+	}) {
+		super({ objectMode: true, highWaterMark: numBuffers - 1 });
+		this.chunkSize = chunkSize;
+		this.alignedReadableState = new AlignedReadableState(
+			chunkSize,
+			alignment,
+			numBuffers,
+		);
 	}
 
-	private writeBuffers(flush = false) {
-		if (flush || this._bytes >= this.chunkSize) {
-			let block = Buffer.concat(this._buffers, this._bytes);
+	private async writeBuffers(flush = false): Promise<void> {
+		if (flush || this.inputBytes >= this.chunkSize) {
+			// TODO optimize
+			let block = Buffer.concat(this.inputBuffers, this.inputBytes);
 			const length = flush
 				? block.length
 				: Math.floor(block.length / this.chunkSize) * this.chunkSize;
 
-			this._buffers.length = 0;
-			this._bytes = 0;
+			this.inputBuffers.length = 0;
+			this.inputBytes = 0;
 
 			if (block.length !== length) {
-				this._buffers.push(block.slice(length));
-				this._bytes += block.length - length;
+				this.inputBuffers.push(block.slice(length));
+				this.inputBytes += block.length - length;
 				block = block.slice(0, length);
 			}
 
-			this.bytesWritten += block.length;
-			this.push(block);
+			const alignedBuffer = this.alignedReadableState.getCurrentBuffer();
+			const unlock = await alignedBuffer.lock();
+			block.copy(alignedBuffer);
+			unlock();
```

Comment:
Should this be
```suggestion
			const unlock = await alignedBuffer.lock();
			try {
				block.copy(alignedBuffer);
			} finally {
				unlock();
			}
```
or can the copy never fail or not unlocking on failure is ok? (same for other cases like this)

Reply:
updated